File: frontend/src/lib/flexibleWeights.ts
export interface FlexibleWeightMapping {
  left: string
  right: string
}

export interface FlexibleStockInput {
  symbol: string
  currentWeightPct: number
  targetWeight: number
  rebalDollars: number
}

export interface FlexibleStockDisplay {
  targetWeight: Record<string, number>
  rebalDollars: Record<string, number>
}

export interface FlexibleWeightRuleLimitCheck {
  key: 'expandedExpressions' | 'expressionTerms'
  label: string
  current: number
  limit: number
  over: boolean
}

export interface FlexibleWeightRuleValidation {
  ok: boolean
  checks: FlexibleWeightRuleLimitCheck[]
}

interface ExpressionTerm {
  symbol: string
  weight: number
}

interface ParsedFlexibleRule {
  leftTerms: ExpressionTerm[]
  rightTerms: ExpressionTerm[]
  leftKey: string
  rightKey: string
}

const EPSILON = 1e-9
const MAX_EXPANDED_EXPRESSIONS = 100
const MAX_EXPRESSION_TERMS = 12
const MAX_RULE_APPLICATION_PASSES = 20

interface ExpansionStats {
  maxExpandedExpressions: number
  maxExpressionTerms: number
  expressionsExceeded: boolean
  termsExceeded: boolean
}

export function parseFlexibleWeightMappings(raw?: string | null): FlexibleWeightMapping[] {
  if (!raw?.trim()) return []
  try {
    const parsed = JSON.parse(raw)
    if (!Array.isArray(parsed)) return []
    return parsed.map(item => {
      const obj = item && typeof item === 'object' ? item as Partial<FlexibleWeightMapping> : {}
      return {
        left: String(obj.left ?? '').trim(),
        right: String(obj.right ?? '').trim(),
      }
    }).filter(row => row.left && row.right)
  } catch {
    return []
  }
}

export function serializeFlexibleWeightMappings(rows: FlexibleWeightMapping[]): string {
  const clean = rows
    .map(row => ({ left: row.left.trim(), right: row.right.trim() }))
    .filter(row => row.left && row.right)
  return clean.length ? JSON.stringify(clean) : ''
}

export function parseFlexibleExpression(raw: string): ExpressionTerm[] {
  const tokens = raw.trim().split(/\s+/).filter(Boolean)
  const terms: ExpressionTerm[] = []
  for (let i = 0; i < tokens.length; i += 1) {
    const maybeWeight = parseFloat(tokens[i])
    if (Number.isFinite(maybeWeight) && i + 1 < tokens.length && !Number.isNaN(Number(tokens[i]))) {
      terms.push({ weight: Math.abs(maybeWeight), symbol: tokens[i + 1].trim().toUpperCase() })
      i += 1
    } else {
      terms.push({ weight: 1, symbol: tokens[i].trim().toUpperCase() })
    }
  }
  return terms.filter(term => term.symbol && term.weight > 0)
}

function normalizeWeight(value: number): number {
  return Math.round(value * 1e10) / 1e10
}

function normalizeTerms(terms: ExpressionTerm[]): ExpressionTerm[] {
  const weights = new Map<string, number>()
  for (const term of terms) {
    const symbol = term.symbol.trim().toUpperCase()
    const weight = normalizeWeight(Math.abs(term.weight))
    if (!symbol || weight <= EPSILON) continue
    weights.set(symbol, normalizeWeight((weights.get(symbol) ?? 0) + weight))
  }
  return [...weights.entries()]
    .map(([symbol, weight]) => ({ symbol, weight }))
    .filter(term => term.weight > EPSILON)
    .sort((a, b) => a.symbol.localeCompare(b.symbol))
}

function expressionKey(terms: ExpressionTerm[]): string {
  return normalizeTerms(terms)
    .map(term => `${term.symbol}:${term.weight}`)
    .join('|')
}

function scaleTerms(terms: ExpressionTerm[], scale: number): ExpressionTerm[] {
  return normalizeTerms(terms.map(term => ({ symbol: term.symbol, weight: term.weight * scale })))
}

function parseFlexibleRule(mapping: FlexibleWeightMapping): ParsedFlexibleRule | null {
  const leftTerms = normalizeTerms(parseFlexibleExpression(mapping.left))
  const rightTerms = normalizeTerms(parseFlexibleExpression(mapping.right))
  if (!leftTerms.length || !rightTerms.length) return null
  return {
    leftTerms,
    rightTerms,
    leftKey: expressionKey(leftTerms),
    rightKey: expressionKey(rightTerms),
  }
}

function uniqueSymbols(terms: ExpressionTerm[]): string[] {
  return [...new Set(terms.map(term => term.symbol))]
}

function expressionUnits(values: Record<string, number>, terms: ExpressionTerm[]): number {
  let sign = 0
  let units = Infinity
  for (const term of terms) {
    const value = values[term.symbol]
    if (value === undefined || value === 0) return 0
    const legUnits = value / term.weight
    const legSign = Math.sign(legUnits)
    if (sign === 0) {
      sign = legSign
    } else if (legSign !== sign) {
      return 0
    }
    units = Math.min(units, Math.abs(legUnits))
  }
  return sign * (Number.isFinite(units) ? units : 0)
}

function applyExpressionShift(values: Record<string, number>, terms: ExpressionTerm[], shiftUnits: number) {
  if (shiftUnits === 0) return
  for (const term of terms) {
    values[term.symbol] = (values[term.symbol] ?? 0) + shiftUnits * term.weight
  }
}

function getFlexibleWeightShift(
  targetWeight: Record<string, number>,
  currentWeight: Record<string, number>,
  leftTerms: ExpressionTerm[],
  rightTerms: ExpressionTerm[],
): number {
  const weightDev = Object.fromEntries(
    uniqueSymbols([...leftTerms, ...rightTerms]).map(symbol => [
      symbol,
      (currentWeight[symbol] ?? 0) - (targetWeight[symbol] ?? 0),
    ]),
  )
  const leftDev = expressionUnits(weightDev, leftTerms)
  const rightDev = expressionUnits(weightDev, rightTerms)

  if (leftDev > 0 && rightDev < 0) {
    return Math.min(leftDev, -rightDev)
  } else if (rightDev > 0 && leftDev < 0) {
    return -Math.min(rightDev, -leftDev)
  }
  return 0
}

function applyFlexibleWeightRule(
  targetWeight: Record<string, number>,
  currentWeight: Record<string, number>,
  leftTerms: ExpressionTerm[],
  rightTerms: ExpressionTerm[],
): boolean {
  const shift = getFlexibleWeightShift(targetWeight, currentWeight, leftTerms, rightTerms)
  if (Math.abs(shift) <= EPSILON) return false
  applyExpressionShift(targetWeight, leftTerms, shift)
  applyExpressionShift(targetWeight, rightTerms, -shift)
  return true
}

function getFlexibleRebalShift(
  rebalDollars: Record<string, number>,
  leftTerms: ExpressionTerm[],
  rightTerms: ExpressionTerm[],
): number {
  const leftRebal = expressionUnits(rebalDollars, leftTerms)
  const rightRebal = expressionUnits(rebalDollars, rightTerms)

  if (leftRebal < 0 && rightRebal > 0) {
    return Math.min(-leftRebal, rightRebal)
  } else if (rightRebal < 0 && leftRebal > 0) {
    return -Math.min(-rightRebal, leftRebal)
  }
  return 0
}

function applyFlexibleRebalRule(
  rebalDollars: Record<string, number>,
  leftTerms: ExpressionTerm[],
  rightTerms: ExpressionTerm[],
): boolean {
  const shift = getFlexibleRebalShift(rebalDollars, leftTerms, rightTerms)
  if (Math.abs(shift) <= EPSILON) return false
  applyExpressionShift(rebalDollars, leftTerms, shift)
  applyExpressionShift(rebalDollars, rightTerms, -shift)
  return true
}

function expressionWithinLimit(terms: ExpressionTerm[]): boolean {
  return terms.length > 0 && terms.length <= MAX_EXPRESSION_TERMS
}

function createExpansionStats(): ExpansionStats {
  return {
    maxExpandedExpressions: 0,
    maxExpressionTerms: 0,
    expressionsExceeded: false,
    termsExceeded: false,
  }
}

function recordExpansionStats(stats: ExpansionStats | undefined, terms: ExpressionTerm[]) {
  if (!stats) return
  stats.maxExpressionTerms = Math.max(stats.maxExpressionTerms, terms.length)
  if (terms.length > MAX_EXPRESSION_TERMS) stats.termsExceeded = true
}

// Composite rules rewrite only when the full expression matches; single-term rules may rewrite one leg inside a larger expression.
function rewriteWholeExpression(terms: ExpressionTerm[], rules: ParsedFlexibleRule[]): ExpressionTerm[][] {
  const key = expressionKey(terms)
  const next: ExpressionTerm[][] = []
  for (const rule of rules) {
    if (key === rule.leftKey) next.push(rule.rightTerms)
    if (key === rule.rightKey) next.push(rule.leftTerms)
  }
  return next
}

function rewriteSingleTerms(terms: ExpressionTerm[], rules: ParsedFlexibleRule[]): ExpressionTerm[][] {
  const next: ExpressionTerm[][] = []
  for (let i = 0; i < terms.length; i += 1) {
    const term = terms[i]
    const rest = terms.filter((_, index) => index !== i)
    for (const rule of rules) {
      if (rule.leftTerms.length === 1 && rule.leftTerms[0].symbol === term.symbol) {
        next.push(normalizeTerms([
          ...rest,
          ...scaleTerms(rule.rightTerms, term.weight / rule.leftTerms[0].weight),
        ]))
      }
      if (rule.rightTerms.length === 1 && rule.rightTerms[0].symbol === term.symbol) {
        next.push(normalizeTerms([
          ...rest,
          ...scaleTerms(rule.leftTerms, term.weight / rule.rightTerms[0].weight),
        ]))
      }
    }
  }
  return next
}

function expandExpression(
  startTerms: ExpressionTerm[],
  rules: ParsedFlexibleRule[],
  stats?: ExpansionStats,
): ExpressionTerm[][] {
  const start = normalizeTerms(startTerms)
  const seen = new Map<string, ExpressionTerm[]>()
  const queue: ExpressionTerm[][] = [start]
  seen.set(expressionKey(start), start)
  recordExpansionStats(stats, start)
  if (stats) stats.maxExpandedExpressions = Math.max(stats.maxExpandedExpressions, seen.size)

  for (let cursor = 0; cursor < queue.length; cursor += 1) {
    const current = queue[cursor]

    const candidates = [
      ...rewriteWholeExpression(current, rules),
      ...rewriteSingleTerms(current, rules),
    ]

    for (const candidate of candidates) {
      const normalized = normalizeTerms(candidate)
      recordExpansionStats(stats, normalized)
      if (!expressionWithinLimit(normalized)) continue
      const key = expressionKey(normalized)
      if (seen.has(key)) continue
      if (seen.size >= MAX_EXPANDED_EXPRESSIONS) {
        if (stats) stats.expressionsExceeded = true
        return [...seen.values()]
      }
      seen.set(key, normalized)
      queue.push(normalized)
      recordExpansionStats(stats, normalized)
      if (stats) stats.maxExpandedExpressions = Math.max(stats.maxExpandedExpressions, seen.size)
    }
  }

  return [...seen.values()]
}

function getExpandedExpression(
  terms: ExpressionTerm[],
  rules: ParsedFlexibleRule[],
  cache: Map<string, ExpressionTerm[][]>,
): ExpressionTerm[][] {
  const key = expressionKey(terms)
  const cached = cache.get(key)
  if (cached) return cached
  const expanded = expandExpression(terms, rules)
  cache.set(key, expanded)
  return expanded
}

function applyBestFlexibleWeightRule(
  targetWeight: Record<string, number>,
  currentWeight: Record<string, number>,
  leftCandidates: ExpressionTerm[][],
  rightCandidates: ExpressionTerm[][],
): boolean {
  let bestLeftTerms: ExpressionTerm[] | null = null
  let bestRightTerms: ExpressionTerm[] | null = null
  let bestShift = 0

  for (const leftTerms of leftCandidates) {
    const leftKey = expressionKey(leftTerms)
    for (const rightTerms of rightCandidates) {
      if (leftKey === expressionKey(rightTerms)) continue
      const shift = getFlexibleWeightShift(targetWeight, currentWeight, leftTerms, rightTerms)
      if (Math.abs(shift) > Math.abs(bestShift) + EPSILON) {
        bestLeftTerms = leftTerms
        bestRightTerms = rightTerms
        bestShift = shift
      }
    }
  }

  if (!bestLeftTerms || !bestRightTerms || Math.abs(bestShift) <= EPSILON) return false
  return applyFlexibleWeightRule(targetWeight, currentWeight, bestLeftTerms, bestRightTerms)
}

function applyBestFlexibleRebalRule(
  rebalDollars: Record<string, number>,
  leftCandidates: ExpressionTerm[][],
  rightCandidates: ExpressionTerm[][],
): boolean {
  let bestLeftTerms: ExpressionTerm[] | null = null
  let bestRightTerms: ExpressionTerm[] | null = null
  let bestShift = 0

  for (const leftTerms of leftCandidates) {
    const leftKey = expressionKey(leftTerms)
    for (const rightTerms of rightCandidates) {
      if (leftKey === expressionKey(rightTerms)) continue
      const shift = getFlexibleRebalShift(rebalDollars, leftTerms, rightTerms)
      if (Math.abs(shift) > Math.abs(bestShift) + EPSILON) {
        bestLeftTerms = leftTerms
        bestRightTerms = rightTerms
        bestShift = shift
      }
    }
  }

  if (!bestLeftTerms || !bestRightTerms || Math.abs(bestShift) <= EPSILON) return false
  return applyFlexibleRebalRule(rebalDollars, bestLeftTerms, bestRightTerms)
}

export function validateFlexibleWeightMappings(mappings: FlexibleWeightMapping[]): FlexibleWeightRuleValidation {
  const parsedRules = mappings
    .map(parseFlexibleRule)
    .filter((rule): rule is ParsedFlexibleRule => rule !== null)
  const stats = createExpansionStats()
  const expandedKeys = new Set<string>()

  for (const rule of parsedRules) {
    for (const terms of [rule.leftTerms, rule.rightTerms]) {
      const key = expressionKey(terms)
      if (expandedKeys.has(key)) continue
      expandedKeys.add(key)
      expandExpression(terms, parsedRules, stats)
    }
  }

  const checks: FlexibleWeightRuleLimitCheck[] = [
    {
      key: 'expandedExpressions',
      label: 'Expanded expressions',
      current: stats.expressionsExceeded ? MAX_EXPANDED_EXPRESSIONS + 1 : stats.maxExpandedExpressions,
      limit: MAX_EXPANDED_EXPRESSIONS,
      over: stats.expressionsExceeded,
    },
    {
      key: 'expressionTerms',
      label: 'Expression terms',
      current: stats.termsExceeded ? Math.max(stats.maxExpressionTerms, MAX_EXPRESSION_TERMS + 1) : stats.maxExpressionTerms,
      limit: MAX_EXPRESSION_TERMS,
      over: stats.termsExceeded,
    },
  ]

  return {
    ok: checks.every(check => !check.over),
    checks,
  }
}

export function computeFlexibleStockDisplay(
  stocks: FlexibleStockInput[],
  mappings: FlexibleWeightMapping[],
): FlexibleStockDisplay {
  const targetWeight = Object.fromEntries(stocks.map(stock => [stock.symbol, stock.targetWeight]))
  const currentWeight = Object.fromEntries(stocks.map(stock => [stock.symbol, stock.currentWeightPct]))
  const rebalDollars = Object.fromEntries(stocks.map(stock => [stock.symbol, stock.rebalDollars]))
  const parsedRules = mappings
    .map(parseFlexibleRule)
    .filter((rule): rule is ParsedFlexibleRule => rule !== null)
  const expansionCache = new Map<string, ExpressionTerm[][]>()

  for (let pass = 0; pass < MAX_RULE_APPLICATION_PASSES; pass += 1) {
    let changed = false
    for (const rule of parsedRules) {
      const leftCandidates = getExpandedExpression(rule.leftTerms, parsedRules, expansionCache)
      const rightCandidates = getExpandedExpression(rule.rightTerms, parsedRules, expansionCache)
      changed = applyBestFlexibleWeightRule(targetWeight, currentWeight, leftCandidates, rightCandidates) || changed
      changed = applyBestFlexibleRebalRule(rebalDollars, leftCandidates, rightCandidates) || changed
    }
    if (!changed) break
  }

  return { targetWeight, rebalDollars }
}

File: android/app/src/main/java/com/portfoliohelper/domain/FlexibleWeights.kt
package com.portfoliohelper.domain

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.round
import kotlin.math.sign

@Serializable
data class FlexibleWeightMapping(val left: String, val right: String)

data class FlexibleStockInput(
    val symbol: String,
    val currentWeightPct: Double,
    val targetWeight: Double,
    val rebalDollars: Double,
)

data class FlexibleStockDisplay(
    val targetWeight: Map<String, Double>,
    val rebalDollars: Map<String, Double>,
)

private data class ExpressionTerm(val symbol: String, val weight: Double)

private data class ParsedFlexibleRule(
    val leftTerms: List<ExpressionTerm>,
    val rightTerms: List<ExpressionTerm>,
    val leftKey: String,
    val rightKey: String,
)

private const val EPSILON = 1e-9
private const val MAX_EXPANDED_EXPRESSIONS = 100
private const val MAX_EXPRESSION_TERMS = 12
private const val MAX_RULE_APPLICATION_PASSES = 20
private val flexibleJson = Json { ignoreUnknownKeys = true }

fun parseFlexibleWeightMappings(raw: String?): List<FlexibleWeightMapping> {
    if (raw.isNullOrBlank()) return emptyList()
    return runCatching {
        flexibleJson.parseToJsonElement(raw).jsonArray.mapNotNull { item ->
            runCatching {
                val fields = item.jsonObject
                FlexibleWeightMapping(
                    fields["left"]?.jsonPrimitive?.contentOrNull.orEmpty().trim(),
                    fields["right"]?.jsonPrimitive?.contentOrNull.orEmpty().trim(),
                )
            }.getOrNull()
        }
            .filter { it.left.isNotBlank() && it.right.isNotBlank() }
    }.getOrDefault(emptyList())
}

fun hasFlexibleWeightMappings(raw: String?): Boolean =
    parseFlexibleWeightMappings(raw).any { parseFlexibleRule(it) != null }

fun computeFlexibleStockDisplay(
    stocks: List<FlexibleStockInput>,
    mappings: List<FlexibleWeightMapping>,
): FlexibleStockDisplay {
    val targetWeight = stocks.associate { it.symbol to it.targetWeight }.toMutableMap()
    val currentWeight = stocks.associate { it.symbol to it.currentWeightPct }
    val rebalDollars = stocks.associate { it.symbol to it.rebalDollars }.toMutableMap()
    val rules = mappings.mapNotNull(::parseFlexibleRule)
    val expansionCache = mutableMapOf<String, List<List<ExpressionTerm>>>()

    repeat(MAX_RULE_APPLICATION_PASSES) {
        var changed = false
        for (rule in rules) {
            val leftCandidates = getExpandedExpression(rule.leftTerms, rules, expansionCache)
            val rightCandidates = getExpandedExpression(rule.rightTerms, rules, expansionCache)
            changed = applyBestFlexibleWeightRule(
                targetWeight,
                currentWeight,
                leftCandidates,
                rightCandidates,
            ) || changed
            changed = applyBestFlexibleRebalRule(
                rebalDollars,
                leftCandidates,
                rightCandidates,
            ) || changed
        }
        if (!changed) return FlexibleStockDisplay(targetWeight, rebalDollars)
    }
    return FlexibleStockDisplay(targetWeight, rebalDollars)
}

private fun parseFlexibleExpression(raw: String): List<ExpressionTerm> {
    val tokens = raw.trim().split(Regex("\\s+")).filter(String::isNotBlank)
    val terms = mutableListOf<ExpressionTerm>()
    var index = 0
    while (index < tokens.size) {
        val weight = tokens[index].toDoubleOrNull()
        if (weight != null && weight.isFinite() && index + 1 < tokens.size) {
            terms += ExpressionTerm(tokens[index + 1].trim().uppercase(), abs(weight))
            index += 2
        } else {
            terms += ExpressionTerm(tokens[index].trim().uppercase(), 1.0)
            index += 1
        }
    }
    return terms.filter { it.symbol.isNotBlank() && it.weight > 0 }
}

private fun normalizeWeight(value: Double): Double = round(value * 1e10) / 1e10

private fun normalizeTerms(terms: List<ExpressionTerm>): List<ExpressionTerm> {
    val weights = mutableMapOf<String, Double>()
    for (term in terms) {
        val symbol = term.symbol.trim().uppercase()
        val weight = normalizeWeight(abs(term.weight))
        if (symbol.isBlank() || weight <= EPSILON) continue
        weights[symbol] = normalizeWeight((weights[symbol] ?: 0.0) + weight)
    }
    return weights.map { ExpressionTerm(it.key, it.value) }
        .filter { it.weight > EPSILON }
        .sortedBy { it.symbol }
}

private fun expressionKey(terms: List<ExpressionTerm>): String = normalizeTerms(terms)
    .joinToString("|") { "${it.symbol}:${it.weight}" }

private fun scaleTerms(terms: List<ExpressionTerm>, scale: Double): List<ExpressionTerm> =
    normalizeTerms(terms.map { it.copy(weight = it.weight * scale) })

private fun parseFlexibleRule(mapping: FlexibleWeightMapping): ParsedFlexibleRule? {
    val left = normalizeTerms(parseFlexibleExpression(mapping.left))
    val right = normalizeTerms(parseFlexibleExpression(mapping.right))
    if (left.isEmpty() || right.isEmpty()) return null
    return ParsedFlexibleRule(left, right, expressionKey(left), expressionKey(right))
}

private fun expressionUnits(values: Map<String, Double>, terms: List<ExpressionTerm>): Double {
    var valueSign = 0.0
    var units = Double.POSITIVE_INFINITY
    for (term in terms) {
        val value = values[term.symbol] ?: return 0.0
        if (value == 0.0) return 0.0
        val legUnits = value / term.weight
        if (valueSign == 0.0) valueSign = legUnits.sign else if (legUnits.sign != valueSign) return 0.0
        units = min(units, abs(legUnits))
    }
    return valueSign * if (units.isFinite()) units else 0.0
}

private fun applyExpressionShift(
    values: MutableMap<String, Double>,
    terms: List<ExpressionTerm>,
    shiftUnits: Double,
) {
    if (shiftUnits == 0.0) return
    terms.forEach { values[it.symbol] = (values[it.symbol] ?: 0.0) + shiftUnits * it.weight }
}

private fun flexibleWeightShift(
    targetWeight: Map<String, Double>,
    currentWeight: Map<String, Double>,
    left: List<ExpressionTerm>,
    right: List<ExpressionTerm>,
): Double {
    val deviations = (left + right).map { it.symbol }.distinct().associateWith {
        (currentWeight[it] ?: 0.0) - (targetWeight[it] ?: 0.0)
    }
    val leftDeviation = expressionUnits(deviations, left)
    val rightDeviation = expressionUnits(deviations, right)
    return when {
        leftDeviation > 0 && rightDeviation < 0 -> min(leftDeviation, -rightDeviation)
        rightDeviation > 0 && leftDeviation < 0 -> -min(rightDeviation, -leftDeviation)
        else -> 0.0
    }
}

private fun flexibleRebalShift(
    rebalDollars: Map<String, Double>,
    left: List<ExpressionTerm>,
    right: List<ExpressionTerm>,
): Double {
    val leftRebal = expressionUnits(rebalDollars, left)
    val rightRebal = expressionUnits(rebalDollars, right)
    return when {
        leftRebal < 0 && rightRebal > 0 -> min(-leftRebal, rightRebal)
        rightRebal < 0 && leftRebal > 0 -> -min(-rightRebal, leftRebal)
        else -> 0.0
    }
}

private fun applyFlexibleWeightRule(
    targetWeight: MutableMap<String, Double>,
    currentWeight: Map<String, Double>,
    left: List<ExpressionTerm>,
    right: List<ExpressionTerm>,
): Boolean {
    val shift = flexibleWeightShift(targetWeight, currentWeight, left, right)
    if (abs(shift) <= EPSILON) return false
    applyExpressionShift(targetWeight, left, shift)
    applyExpressionShift(targetWeight, right, -shift)
    return true
}

private fun applyFlexibleRebalRule(
    rebalDollars: MutableMap<String, Double>,
    left: List<ExpressionTerm>,
    right: List<ExpressionTerm>,
): Boolean {
    val shift = flexibleRebalShift(rebalDollars, left, right)
    if (abs(shift) <= EPSILON) return false
    applyExpressionShift(rebalDollars, left, shift)
    applyExpressionShift(rebalDollars, right, -shift)
    return true
}

private fun rewriteWholeExpression(
    terms: List<ExpressionTerm>,
    rules: List<ParsedFlexibleRule>,
): List<List<ExpressionTerm>> = buildList {
    val key = expressionKey(terms)
    rules.forEach { rule ->
        if (key == rule.leftKey) add(rule.rightTerms)
        if (key == rule.rightKey) add(rule.leftTerms)
    }
}

private fun rewriteSingleTerms(
    terms: List<ExpressionTerm>,
    rules: List<ParsedFlexibleRule>,
): List<List<ExpressionTerm>> = buildList {
    terms.forEachIndexed { index, term ->
        val rest = terms.filterIndexed { candidate, _ -> candidate != index }
        rules.forEach { rule ->
            if (rule.leftTerms.size == 1 && rule.leftTerms.single().symbol == term.symbol) {
                add(normalizeTerms(rest + scaleTerms(rule.rightTerms, term.weight / rule.leftTerms.single().weight)))
            }
            if (rule.rightTerms.size == 1 && rule.rightTerms.single().symbol == term.symbol) {
                add(normalizeTerms(rest + scaleTerms(rule.leftTerms, term.weight / rule.rightTerms.single().weight)))
            }
        }
    }
}

private fun expandExpression(
    startTerms: List<ExpressionTerm>,
    rules: List<ParsedFlexibleRule>,
): List<List<ExpressionTerm>> {
    val start = normalizeTerms(startTerms)
    val seen = linkedMapOf(expressionKey(start) to start)
    val queue = mutableListOf(start)
    var cursor = 0
    while (cursor < queue.size) {
        val current = queue[cursor++]
        val candidates = rewriteWholeExpression(current, rules) + rewriteSingleTerms(current, rules)
        for (candidate in candidates) {
            val normalized = normalizeTerms(candidate)
            if (normalized.isEmpty() || normalized.size > MAX_EXPRESSION_TERMS) continue
            val key = expressionKey(normalized)
            if (key in seen) continue
            if (seen.size >= MAX_EXPANDED_EXPRESSIONS) return seen.values.toList()
            seen[key] = normalized
            queue += normalized
        }
    }
    return seen.values.toList()
}

private fun getExpandedExpression(
    terms: List<ExpressionTerm>,
    rules: List<ParsedFlexibleRule>,
    cache: MutableMap<String, List<List<ExpressionTerm>>>,
): List<List<ExpressionTerm>> = cache.getOrPut(expressionKey(terms)) { expandExpression(terms, rules) }

private fun applyBestFlexibleWeightRule(
    targetWeight: MutableMap<String, Double>,
    currentWeight: Map<String, Double>,
    leftCandidates: List<List<ExpressionTerm>>,
    rightCandidates: List<List<ExpressionTerm>>,
): Boolean {
    var bestLeft: List<ExpressionTerm>? = null
    var bestRight: List<ExpressionTerm>? = null
    var bestShift = 0.0
    for (left in leftCandidates) for (right in rightCandidates) {
        if (expressionKey(left) == expressionKey(right)) continue
        val shift = flexibleWeightShift(targetWeight, currentWeight, left, right)
        if (abs(shift) > abs(bestShift) + EPSILON) {
            bestLeft = left
            bestRight = right
            bestShift = shift
        }
    }
    return if (bestLeft != null && bestRight != null && abs(bestShift) > EPSILON) {
        applyFlexibleWeightRule(targetWeight, currentWeight, bestLeft, bestRight)
    } else false
}

private fun applyBestFlexibleRebalRule(
    rebalDollars: MutableMap<String, Double>,
    leftCandidates: List<List<ExpressionTerm>>,
    rightCandidates: List<List<ExpressionTerm>>,
): Boolean {
    var bestLeft: List<ExpressionTerm>? = null
    var bestRight: List<ExpressionTerm>? = null
    var bestShift = 0.0
    for (left in leftCandidates) for (right in rightCandidates) {
        if (expressionKey(left) == expressionKey(right)) continue
        val shift = flexibleRebalShift(rebalDollars, left, right)
        if (abs(shift) > abs(bestShift) + EPSILON) {
            bestLeft = left
            bestRight = right
            bestShift = shift
        }
    }
    return if (bestLeft != null && bestRight != null && abs(bestShift) > EPSILON) {
        applyFlexibleRebalRule(rebalDollars, bestLeft, bestRight)
    } else false
}
